cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc

```diff
@@ -769,6 +770,41 @@ std::vector<std::shared_ptr<CastFunction>> GetNumericCasts() {
   return functions;
 }
 
+
+Status CastToExtension(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+  const CastOptions& options = checked_cast<const CastState*>(ctx->state())->options;
+
+  DCHECK(batch[0].is_array());
+  std::shared_ptr<Array> array = batch[0].array.ToArray();
+  std::shared_ptr<Array> result;
+  auto ext_ty = GetExtensionType(kOutputTargetType.type()->name());
+  if (ext_ty == nullptr) {
+    return Status::Invalid("Could not find extension type: " + kOutputTargetType.type()->name());
+  }
+  auto out_ty = ext_ty->storage_type();
+  RETURN_NOT_OK(Cast(*array, out_ty, options,
+                     ctx->exec_context())
+                    .Value(&result));
+  ExtensionArray extension(ext_ty, result);
+  out->value = std::move(extension.data());
+  return Status::OK();
+}
+
+std::shared_ptr<CastFunction> GetCastToExtension(std::string name) {
+  auto func = std::make_shared<CastFunction>(std::move(name), Type::EXTENSION);
+  auto out_ty = kOutputTargetType.type();
```

Comment:
@rok I feel like I've exhausted my debugging options here. While I know these functions are in the wrong file.. was trying to get a working prototype.

However, this line here aborts:
```python
In [1]: import pyarrow as pa

In [2]: from pyarrow.tests.test_extension_type import UuidType, IntegerType

In [3]: a = pa.array([1, 2, 3], pa.int64())

In [4]: a.cast(IntegerType())
/home/milesg/Projects/arrow/cpp/src/arrow/compute/kernel.cc:382:  Check failed: (FIXED) == (kind_)
```
Do you have any suggestions?

Reply:
See my comment below.
For printing - you can also `#include "arrow/util/logging.h"` and then do:
```
ARROW_LOG(INFO) << "text" << value;
```

Reply:
Handy, thank you!